src/gc.c

```diff
@@ -2317,7 +2430,53 @@ JL_DLLEXPORT void jl_gc_collect(int full)
 #endif
     if (recollect) {
         n_pause--;
-        jl_gc_collect(0);
+        _jl_gc_collect(0, stack_hi);
+    }
+}
+
+JL_DLLEXPORT void jl_gc_collect(int full)
+{
+    if (jl_gc_disable_counter)
+        return;
+    char *stack_hi = (char*)gc_get_stack_ptr();
+    gc_debug_print();
+    JL_SIGATOMIC_BEGIN();
+
+    int8_t old_state = jl_get_ptls_states()->gc_state;
+    jl_get_ptls_states()->gc_state = 1;
+    // In case multiple threads enter the GC at the same time, only allow
+    // one of them to actually run the collection. We can't just let the
+    // master thread do the GC since it might be running unmanaged code
+    // and can take arbitrarily long time before hitting a safe point.
+    if (!JL_ATOMIC_COMPARE_AND_SWAP(jl_gc_running, 0, 1)) {
+#ifdef JULIA_ENABLE_THREADING
+        JL_SIGATOMIC_END();
+        jl_wait_for_gc();
+        jl_gc_state_set(old_state, 1);
+#else
+        // For single thread, GC should not call itself (in finalizers) before
+        // setting jl_gc_running to false so this should never happen.
+        assert(0 && "GC synchronization failure");
```

Comment:
It's during bootstrap
```
make[1]: Entering directory `/d/code/msys64/home/Tony/julia'
 cd /d/code/msys64/home/Tony/julia/base && /d/code/msys64/home/Tony/julia/usr/bi
n/julia.exe -C x86-64 --output-ji /d/code/msys64/home/Tony/julia/usr/lib/julia/i
nference0.ji -f coreimg.jl
essentials.jl
reflection.jl
options.jl
promotion.jl
tuple.jl
range.jl
expr.jl
error.jl
bool.jl
number.jl
int.jl
operators.jl
pointer.jl
abstractarray.jl
array.jl
hashing.jl
nofloat_hashing.jl
functors.jl
reduce.jl
intset.jl
dict.jl
iterator.jl
inference.jl
Assertion failed: 0 && "GC synchronization failure", file d:/code/msys64/home/To
ny/julia/src/gc.c, line 2459
make[1]: *** [/d/code/msys64/home/Tony/julia/usr/lib/julia/inference0.ji] Error
3
make[1]: Leaving directory `/d/code/msys64/home/Tony/julia'
make: *** [julia-inference] Error 2
```
not sure about getting a backtrace in an msvc build, requires windbg